File: 2023/go/day07.go
package main

import (
	"cmp"
	"fmt"
	"math"
	"slices"
	"strconv"
	"strings"
	"time"
)

type Hand struct {
	hand string
	bid, hand_value int
	card_value float64
}

func hand_value(hand string, with_jokers bool) int {
	counter := make(map[rune]int)
	for _, card := range hand {
		counter[card]++
	}

	n_jokers := 0
	if with_jokers {
		n_jokers = counter['J']
	}

	card_occurences := []int{0, 0}
	for card, count := range counter {
		if with_jokers && card == 'J' {
			continue
		}
		card_occurences = append(card_occurences, count)
	}

	slices.SortFunc(card_occurences, func(i, j int) int {
		return -cmp.Compare(i, j)
	})

	if card_occurences[0] + n_jokers >= 5 {
		return 6
	}
	if card_occurences[0] + n_jokers >= 4 {
		return 5
	}
	if card_occurences[0] + n_jokers >= 3 {
		rem := 3 - card_occurences[0] - n_jokers
		if card_occurences[1] + rem >= 2 {
			return 4
		}
		return 3
	}
	if card_occurences[0] + n_jokers >= 2 {
		rem := 2 - card_occurences[0] - n_jokers
		if card_occurences[1] + rem >= 2 {
			return 2
		}
		return 1
	}
	return 0
}

// Encode card value in "base 1/13" (i.e. 13 cards)
// E.g. assuming no jokers "T24" would translate to
// 8 * 1/13 + 0 * 1/13^2 + 2* 1/13^3
// 2 = 0th card, T = 8th card, 4 = 2nd card
func card_value(hand string, with_jokers bool) float64 {
	var card_order string
	if with_jokers {
		card_order = "J23456789TQKA"
	} else {
		card_order = "23456789TJQKA"
	}

	card_map := make(map[rune]float64)
	for i, card := range card_order {
		card_map[card] = float64(i)
	}

	base := 1. / float64(len(card_order))
	res := 0.
	for i, card := range hand {
		res += card_map[card] * math.Pow(base, float64(i+1))
	}

	return res
}

func build_hand(hand string, bid int, with_jokers bool) Hand {
	return Hand{
		bid: bid,
		hand_value: hand_value(hand, with_jokers),
		card_value: card_value(hand, with_jokers),
		hand: hand,
	}
}

func compare_hands(h1, h2 Hand) int {
	if h1.hand_value == h2.hand_value {
		return cmp.Compare(h1.card_value, h2.card_value)
	}
	return cmp.Compare(h1.hand_value, h2.hand_value)
}

func day07_parsers(input string, with_jokers bool) int {
	hands := []Hand{}
	for _, line := range strings.Split(input, "\n") {
		fields := strings.Fields(line)
		hand := fields[0]
		bid, err := strconv.Atoi(fields[1])
		if err != nil {
			panic(err)
		}
		hands = append(hands, build_hand(hand, bid, with_jokers))
	}
	slices.SortFunc(hands, compare_hands)

	res := 0
	for i, hand := range hands {
		res += (i+1) * hand.bid
	}
	return res
}


func day07_solution(input string) {
	input = strings.TrimSpace(input)

	start := time.Now()
	fmt.Printf("Task 01: %d\n", day07_parsers(input, false))
	elapsed := time.Since(start)
	fmt.Printf("Time: %s\n", elapsed)

	start = time.Now()
	fmt.Printf("Task 02: %d\n", day07_parsers(input, true))
	elapsed = time.Since(start)
	fmt.Printf("Time: %s\n", elapsed)
}

File: 2023/python/day07.py
from collections import Counter
from pathlib import Path

from loguru import logger
from typer import Typer
from utils import timer

main = Typer()

# Encode card values as hex
# Allows lexicographic sorting on the string
# [-1] is necessary to remove "0x" from e.g. hex(0) == "0x0"
CARD_VALUES = {card: hex(i)[-1] for i, card in enumerate("23456789TJQKA")}
CARD_VALUES_WITH_JOKERS = {card: hex(i)[-1] for i, card in enumerate("J23456789TQKA")}


class Hand:
    """Represent a camel card hand"""

    def __init__(self, hand: str, bid: int, with_jokers: bool = False) -> None:
        """Initialize hand

        hand and bid are saved. The hand is also transformed into a
        hex representation for lexicographic sorting and the card types
        are counted, treating jokers differently if necessary.

        Args:
            hand: the hand
            bid: the bid
            with_jokers: If true, "J" is treated as joker, otherwise as jack. Defaults to False.
        """
        self.hand = hand
        self.bid = bid

        # Count card types separating jokers if necessary
        # Also transform card values to hex
        counts = Counter(hand)
        if with_jokers:
            self.card_values = "".join(CARD_VALUES_WITH_JOKERS[card] for card in hand)
            self.jokers = counts["J"]
            del counts["J"]
        else:
            self.card_values = "".join(CARD_VALUES[card] for card in hand)
            self.jokers = 0

        self.counts = sorted(counts.values(), reverse=True)
        self.type_value = self._type_value()

    def _type_value(self) -> int:
        """Return the type value of the hand

        Jokers are taking into account

        5 cards: 6
        4 cards: 5
        Full house: 4
        3 cards: 3
        2 pairs: 2
        1 pair: 1
        high card: 0
        """
        # Buffer for a hand with only jokers
        # or a hand with only jokers and one other card type
        counts = self.counts + [0] * 2

        if counts[0] + self.jokers == 5:
            return 6

        if counts[0] + self.jokers == 4:
            return 5

        if counts[0] + self.jokers >= 3:
            rem = 3 - counts[0] - self.jokers
            if counts[1] + rem >= 2:
                return 4
            return 3

        if counts[0] + self.jokers >= 2:
            rem = 2 - counts[0] - self.jokers
            if counts[1] + rem >= 2:
                return 2

            return 1

        return 0

    def __repr__(self) -> str:
        """Print hand for debugging."""
        return f"{self.hand} {self.bid}"

    def __lt__(self, other: "Hand") -> bool:
        # If both hands have the same type, we have to compare
        # card values lexicographically
        if self.type_value == other.type_value:
            return self.card_values < other.card_values

        return self.type_value < other.type_value


@timer
def tasks(input: str, with_joker: bool = False) -> int:
    """Solution for tasks 01 and 02

    Args:
        input: Input string
        with_joker: If true treat J as joker, else as jack. Defaults to False.

    Returns:
        Sum over rank*bid of each hand
    """
    hands = []
    for line in input.splitlines():
        hand, bid = line.split()
        hands.append(Hand(hand, int(bid), with_jokers=with_joker))

    return sum(rank * hand.bid for rank, hand in enumerate(sorted(hands), 1))


@main.command()
def entrypoint(path: Path):
    """CLI entrypoint"""
    with open(path, "r") as f:
        input = f.read().strip()

    logger.info(f"Task 01: {tasks(input)}")
    logger.info(f"Task 02: {tasks(input, with_joker=True)}")


if __name__ == "__main__":
    main()
